crates/bellpepper-ed25519/src/circuit.rs

```diff
@@ -0,0 +1,645 @@
+                    &mut cs.namespace(|| {
+                        format!("sum of {}*output and {j}*base", 1 << num_remaining_bits)
+                    }),
+                    &output,
+                    &lookup_table[j],
+                )?;
+                final_lookup_table.push(tmp);
+            }
+
+            window_bits.clear();
+            for j in 0..num_remaining_bits {
+                window_bits.push(scalar[j].clone())
+            }
+
+            output = Self::conditionally_select(
+                &mut cs.namespace(|| "final doubling of output"),
+                &final_lookup_table,
+                &window_bits,
+            )?;
+        }
+
+        Ok(output)
+    }
+
+    pub fn ed25519_scalar_multiplication<CS>(
+        &self,
+        cs: &mut CS,
+        scalar: Vec<Boolean>,
+    ) -> Result<Self, SynthesisError>
+    where
+        CS: ConstraintSystem<F>,
+    {
+        Self::ed25519_scalar_multiplication_windowed(
+            &self,
+            &mut cs.namespace(|| {
+                format!("scalar multiplication with window {DEFAULT_SCALAR_MULT_WINDOW_SIZE}")
+            }),
+            scalar,
+            DEFAULT_SCALAR_MULT_WINDOW_SIZE,
+        )
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use crate::curve::Ed25519Curve;
+
+    use super::*;
+    use bellpepper_core::test_cs::TestConstraintSystem;
+    use num_bigint::{BigUint, RandBigInt};
+    use num_integer::Integer;
+    use num_traits::Zero;
+    use pasta_curves::Fp;
+
+    #[test]
+    fn alloc_affine_point_addition() {
+        let b = Ed25519Curve::basepoint();
+        let mut rng = rand::thread_rng();
+        let scalar = rng.gen_biguint_range(&BigUint::zero(), &Ed25519Curve::order());
+        let p = Ed25519Curve::scalar_multiplication(&b, &scalar);
+        let scalar = rng.gen_biguint_range(&BigUint::zero(), &Ed25519Curve::order());
+        let q = Ed25519Curve::scalar_multiplication(&b, &scalar);
+        let sum_expected_value = &p + &q;
+
+        let mut cs = TestConstraintSystem::<Fp>::new();
+
+        let p_alloc =
+            AllocatedAffinePoint::alloc_affine_point(&mut cs.namespace(|| "alloc point p"), &p);
+        assert!(p_alloc.is_ok());
+        let p_al = p_alloc.unwrap();
+
+        let q_alloc =
+            AllocatedAffinePoint::alloc_affine_point(&mut cs.namespace(|| "alloc point q"), &q);
+        assert!(q_alloc.is_ok());
+        let q_al = q_alloc.unwrap();
+
+        let sum_alloc = AllocatedAffinePoint::ed25519_point_addition(
+            &mut cs.namespace(|| "adding p and q"),
+            &p_al,
+            &q_al,
+        );
+        assert!(sum_alloc.is_ok());
+        let sum_al = sum_alloc.unwrap();
+
+        assert_eq!(sum_expected_value, sum_al.value);
+
+        if !cs.is_satisfied() {
+            eprintln!("{:?}", cs.which_is_unsatisfied())
+        }
+        assert!(cs.is_satisfied());
+        println!("Num constraints = {:?}", cs.num_constraints());
```

Comment:
Suggestion: use `assert_eq` to validate the number of constraints. This is happening in other places of the code base and turned out to be useful way to inform the code reader.

Reply:
Will do in all such locations. Thanks.

Reply:
@emmorais I have replaced the println's with assert_eq in all locations. Please let me know if you have other comments/suggestions.